pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTMethodOrConstructorDeclaration.java

```diff
@@ -19,4 +19,10 @@
     @Override
     JavaOperationSignature getSignature();
 
+
+    /**
+     * Returns the formal parameters node of this method or constructor.
+     */
+    ASTFormalParameters getFormalParameters();
```

Comment:
this is a breaking API change. We should comment it out until 7.0.0, and let the method / constructor declarations have the method on it's own... given the current usage it should work

Reply:
Ah, I see - because it's an interface. I can change it.

However, since PMD provides the AST classes and their implementations - do we really consider this as an API (as in: you can extend it and create your own sub-classes)? It would be very difficult to have PMD actually use this class - since the parser wouldn't return this custom subclass...

Reply:
I think at some point we need to clear-cut which APIs are public and which are not (ie: by using `internal` packages [as Gradle does](https://github.com/gradle/gradle/tree/master/subprojects/core/src/test/groovy/org/gradle/internal)). Up until then, developers may assume anything is at their disposition.

As for these particular interface, I can think of it being used in unit tests, as part of phony AST nodes. Something we actually do ourselves on several ocasions.

Reply:
> we need to clear-cut which APIs are public and which are not (ie: by using internal packages as Gradle does).

I totally support this. [RxJava](https://github.com/ReactiveX/RxJava#versioning) also uses some specific annotations (`@Experimental`, `@Beta`) which I think are a good idea. Maybe we could even have a `@ReservedSubclassing` one, to express that a class or interface is not meant to be subclassed by external clients, and that its inheritance-specific members are private API, while public members are still public API. Which would be useful for all AST related interfaces for example. 

And maybe, concrete AST node classes should have a package-private constructor, and be final. They're not meant to be instantiated by hand, but that could break some tests and rules

Reply:
I've changed it now on master. Clearly defining the API PMD provides, makes sense and will help, if we at some time in the future want to modularize PMD using java9. That would prevent API leaking then at compile time already...
btw. I also updated #881